Graph.java

```diff
@@ -0,0 +1,166 @@
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.LinkedList;
+
+/**
+ * Created by basescu on 06.06.17.
+ */
+
+class Node {
+    public int key;     // letter index in initial matrix: for M[i][j], we have pos = i * cols + j
+    public char value;
+
+    public Node(int key, char value) {
+        this.key = key;
+        this.value = value;
+    }
+
+    public String toString() {
+        return "key = " + key + " value = " + value;
+    }
+}
+
+/*
+Each letter is a node. The graph has edges for each diagonal & straight traversal.
+Finding words means a graph traversal with word prunning. I chose DFS to
+potentially find complete words faster.
+ */
+public class Graph {
+
+    public final HashMap<Integer, Node> intToNode;
+    public final HashMap<Node, LinkedList<Node>> edges;
+    public final int nrNodes;
+
+    public Graph(int rows, int cols, char[][] input) {
+        intToNode = new HashMap<>();
+        edges = new HashMap<>();
+        nrNodes = rows * cols;
+        for (int i = 0 ; i < rows ; i++) {
+            for (int j = 0 ; j < cols ; j++) {
+                int nodeKey = i * cols + j;
+                Node n = fetchOrCreateNode(nodeKey, input[i][j]);
+                LinkedList<Node> neigh = identifyNeighbors(input, i, j, rows, cols);
+                edges.put(n, neigh);
+            }
+        }
+    }
+
+    /* Build words using depth-first search and prune search using word prefixes. */
+    private void createWords(Node lastLetter, StringBuilder sb, HashSet visited, HashSet result) {
+
+        // debug
+        /*
+        System.out.println("last letter "+ lastLetter);
+        System.out.println("visited "+ visited);
+        System.out.println("word so far "+ sb.toString());
+        System.out.println("result so far "+ result);
+        */
+
+        for (Node neigh : edges.get(lastLetter)) {
+            if (!visited.contains(neigh)) {
+                String potentialWord = sb.toString() + neigh.value;
+                if (Dictionary.isWord(potentialWord)) {
+                    result.add(sb.toString() + neigh.value);
+                }
+                if (Dictionary.isPrefix(potentialWord)) {
+                    visited.add(neigh);
+                    sb.append(neigh.value);
+                    createWords(neigh, sb, visited, result);
+                    visited.remove(neigh);
+                    sb.deleteCharAt(sb.length()-1);
+                }
+            }
+        }
+    }
+
+    /* Initialize word creation from each node, if the dictioanry contains a prefix for that letter */
+    public HashSet createWords() {
+        HashSet visited = new HashSet();
```

Comment:
No type safety again... Please always use generic parameters to types. Java compiler probably gives you a warning about it. It is a good practice to always fix compiler warnings, even if they seem silly. Compilers are really good at pointing out potential problems.